src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs

```diff
@@ -468,7 +485,15 @@ int IDebugPendingBreakpoint2.Enable(int fEnable)
                     _engine.DebuggedProcess.WorkerThread.RunOperation(() =>
                     {
                         _engine.DebuggedProcess.AddInternalBreakAction(
-                            () => bp.EnableAsync(_enabled, _engine.DebuggedProcess)
+                            async () => {
+                                try
+                                {
+                                    await bp.EnableAsync(_enabled, _engine.DebuggedProcess);
+                                } catch (UnexpectedMIResultException exc)
+                                {
+                                    this.SetError(new AD7ErrorBreakpoint(this, exc.MIError, enum_BP_ERROR_TYPE.BPET_GENERAL_ERROR), true);
```

Comment:
`-break-enable` can fail if this would exceed the breakpoint limit. The exception is caught and turned into a breakpoint error event, as you can see in this log:

```
1: (262805) <-1024-break-enable 3
1: (262813) ->1024^error,msg="Hardware breakpoints used exceeds limit."
1: (262816) 1024: elapsed time 11
1: (262816) ->(gdb)
1: (262852) ->&"\n"
1: (262877) ->^done
1: (262880) ->(gdb)
1: (262981) Send Event AD7BreakpointErrorEvent
```

However, the VS debugger does not reflect the error. Is this a UI issue?
![vs_screen_enable](https://user-images.githubusercontent.com/34582775/111684167-990ca200-87e3-11eb-8d4c-db1fc33b9498.png)

Reply:
Ping on this one @gregg-miskelly. Is this something that can be addressed in this PR, or does it need investigation in VS debugger UI?

Reply:
Sorry, I missed your question. It isn't obvious to me from reading the code why that wouldn't work. But I do see several suspect things that MIEngine does with error breakpoints:

* IDebugPendingBreakpoint2.CanBind would still return S_OK/no error after this failure (there is no check for existing errors)
* `_enabled` is left true after this failure.
* The `&& ((bpErrorType & enum_BP_ERROR_TYPE.BPET_TYPE_ERROR) != 0)` check in IDebugPendingBreakpoint2.EnumErrorBreakpoints looks wrong to me. I don't know if that has anything to do with this, but I don't think it should exist.


---
In reply to: [602621718](https://github.com/microsoft/MIEngine/pull/1123#discussion_r602621718) [](ancestors = 602621718)

Reply:
Also, we are not failing this call, which I *think* we want to.

Reply:
Thanks for you help. Using your pointers, I was able to get a better experience working. Trying to enable too many breakpoints will pop up an error message and leave the breakpoint unbound.
![image](https://user-images.githubusercontent.com/34582775/113068639-86935080-9173-11eb-9236-488f68d8c294.png)
This required setting `_enabled` to false and making sure to return `E_FAIL` in `Enable`. It also required adding timeout logic similar to what's used in `Bind` to cope with the fact that the `IDebugPendingBreakpoint2` API is synchronous.  I did not end up modifying `CanBind` or `EnumErrorBreakpoints` because neither code path is exercised by the enable logic. Do you still want me to delete that check for the breakpoint error type even though it's unrelated?